python/hidet/distributed/partition/shard.py

```diff
@@ -0,0 +1,237 @@
+    tensor, and 'mesh axis' refers to the axis of hardware hierarchy.
+
+    The constructor can be called by specifying the dimension sharded by each mesh axis. The
+    parameter `sharded_dims` is a list of integers, and the i-th element of it is the dimension
+    sharded by the i-th mesh axis. If any mesh axis does not shard any dimension (tensor are
+    duplicated along that mesh axis), its corresponding value is None.  We assume that each mesh
+    axis can shard at most one dimension, and each dimension can be sharded by multiple mesh axes.
+
+    It also accept a single-value parameter, which will be viewed as a single-layer mesh hierarchy.
+
+    None or negative numbers will be treated as not to shard on this dimension
+    """
+
+    def __init__(self, ndim: int, sharded_dim_per_axis: Union[Optional[int], Sequence[Optional[int]]] = None):
+        self.ndim: int = ndim
+        if not isinstance(sharded_dim_per_axis, Sequence):
+            sharded_dim_per_axis = [sharded_dim_per_axis]
+        for i in range(len(sharded_dim_per_axis)):
+            if sharded_dim_per_axis[i] is not None and sharded_dim_per_axis[i] < 0:
+                sharded_dim_per_axis[i] = None
+        self.sharded_dim_per_axis: List[Optional[int]] = list(sharded_dim_per_axis)
+        self.num_mesh_axis: int = len(sharded_dim_per_axis)
+
+        # Sanity check and build mesh_axes to dim map
+        self.mesh_axes_per_dim: List[List[int]] = [[] for _ in range(self.ndim)]
+        for mesh_axis, dim in enumerate(sharded_dim_per_axis):
+            if dim is not None:
+                if dim >= self.ndim:
+                    raise ValueError("Sharded dim should be less than the number of dimensions")
+                self.mesh_axes_per_dim[dim].append(mesh_axis)
+
+    def is_single_layer_hierarchy(self) -> bool:
+        return self.num_mesh_axis == 1
+
+    def sharded_dim(self) -> int:
+        if not self.is_single_layer_hierarchy():
+            raise ValueError("Only single layer hierarchy supports directly querying the sharded axis.")
+        return self.sharded_dim_per_axis[0]
+
+    def __str__(self) -> str:
+        str_per_dim = [
+            'R' if len(axes) == 0 else ''.join([f'S{axis}' for axis in axes]) for axes in self.mesh_axes_per_dim
+        ]
+        return f"({', '.join(str_per_dim)})"
+
+    def is_full(self) -> bool:
+        return all((len(mesh_axes) == 0 for mesh_axes in self.mesh_axes_per_dim))
+
+    def __eq__(self, other) -> bool:
+        assert isinstance(other, TensorShardSpec)
+        if self.ndim != other.ndim:
+            return False
+        return self._sharded_dim == other._sharded_dim
+
+
+def get_tile(
+    tensor: Union[Tensor, TensorNode], shard_dim: int, nshards: int, rank: int = 0
+) -> Union[Tensor, TensorNode]:
+    # Only works for 1-D partition
+    slice_list = []
+    for i in range(len(tensor.shape)):
+        if i != shard_dim:
```

Comment:
regarding the -1 vs None, I think the code would be cleaner with None meaning not sharded and avoid -1. For example, here,  we can just check if shard_dim is None and be more explicit.